clients/pkg/logentry/stages/limit.go

```diff
@@ -34,36 +42,61 @@ func newLimitStage(logger log.Logger, config interface{}, registerer prometheus.
 		return nil, err
 	}
 
+	logger = log.With(logger, "component", "stage", "type", "limit")
+	if cfg.ByLabelName != "" && cfg.MaxDistinctLabels < MinReasonableMaxDistinctLabels {
+		level.Warn(logger).Log(
+			"msg",
+			fmt.Sprintf("max_distinct_labels was adjusted up to the minimal reasonable value of %d", MinReasonableMaxDistinctLabels),
+		)
+		cfg.MaxDistinctLabels = MinReasonableMaxDistinctLabels
+	}
+
 	r := &limitStage{
-		logger:      log.With(logger, "component", "stage", "type", "limit"),
-		cfg:         cfg,
-		dropCount:   getDropCountMetric(registerer),
-		rateLimiter: rate.NewLimiter(rate.Limit(cfg.Rate), cfg.Burst),
+		logger:    logger,
+		cfg:       cfg,
+		dropCount: getDropCountMetric(registerer),
+	}
+
+	if cfg.ByLabelName != "" {
+		r.dropCountByLabel = getDropCountByLabelMetric(registerer)
+		newRateLimiter := func() *rate.Limiter { return rate.NewLimiter(rate.Limit(cfg.Rate), cfg.Burst) }
+		gcCb := func() { r.dropCountByLabel.Reset() }
+		r.rateLimiterByLabel = util.NewGenMap[model.LabelValue, *rate.Limiter](cfg.MaxDistinctLabels, newRateLimiter, gcCb)
```

Comment:
> for each label value, we will have a separate instance of *rate.Limiter . it will blow up our memory in case we have a lot of instances of stages... 

We don't instantiate Limiter stages dynamically here, stage instances are mapped 1:1 to what defined in the config.  Each stage has fixed `label_name` it cares about and that name is configured by user statically, so there should be no cardinality problem.

Reply:
ok, i see.